### PR TITLE
Add rate limiting to CTIA

### PR DESCRIPTION
**Pull request by msprunck**
Add a rate limiter

> **Epic** threatgrid/iroh#2807

This PR adds a rate limiter by group on all routes. See the modified Readme.md for more details.

When the rate limit is not reached, the following headers are returned in the response:
```
"x-ratelimit-group-limit": "18000",
```

If the rate limit is exceeded:
- The client receives a response with the 429 HTTP status, a `retry-after` header and the JSON message `{"error": "Too Many Requests"}`
- A message is logged with the `:info` level

<a name="qa">[§](#qa)</a> QA
============================

1. Check that Cisco Threat Response client is not rate limited (by checking the response headers)
2. Create an Oauth2 client with CTR and use it with CTIA APIs
3. The response headers `x-ratelimit-group-limit` should be set in the response
4. Make the number of requests to exceed the rate limit and check that future calls are blocked

<a name="ops">[§](#ops)</a> Ops
===============================

- Config change needed: https://github.com/threatgrid/tenzin-config/pull/36

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
public: AMP Global Intel and Private AMP Global Intel APIs are now rate limited
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -318,6 +318,45 @@ use a combination of `offset` and `limit` parameters to paginate results.
 To be used when the result window is superior to `10 000`, allows to easily loop across all pages of a query response.
 use `limit` and `offset` along with `search_after` filled with the value from the `X-Sort` response header to get the next page.
 
+#### Rate limit
+
+Requests may be rate limited by enabling the middleware using the `ctia.http.rate-limit.enabled` property.
+
+It rate limits how many HTTP requests a CTIA group can make in an hour. The group is identified with the property :identity of the current Ring request.
+
+Before the rate limit is reached, the header `X-Ratelimit-Group-Limit` is returned in the response:
```
Comment (ereteog) [inline README.md]: I would also specify that `X-RateLimit-GROUP-Remaining` is also returned.

File: test/ctia/http/middleware/ratelimit_test.clj [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                                       "Origin" "http://external.cisco.com"})]
+      (testing "Group limit"
+        (apply-fixtures
+         ["ctia.http.rate-limit.default-group-limit" 5
+          "ctia.http.rate-limit.enabled" true]
+         (fn []
+           (let [response (call)]
+             (testing "Rate limit headers"
+               (is (= 200 (:status response)))
+               (is (= "5" (get-in response [:headers "X-RateLimit-GROUP-Limit"]))))
```
Comment (ereteog) [inline test/ctia/http/middleware/ratelimit_test.clj]: if `X-RateLimit-GROUP-Remaining` is also in header I would test its presence here.

File: src/ctia/http/middleware/ratelimit.clj [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  (sort-by (juxt second first) #(compare %2 %1) group-limits))
+
+(s/defn ^:always-validate group-limit-fn :- LimitFunction
+  "Builds the group limit function"
+  [{:keys [default-group-limit custom-group-limits]}]
+  (let [group-limits (parse-group-limits custom-group-limits)]
+    (fn [request]
+      (when-let [[group limit]
+                 (some-> (:identity request)
+                         (auth/groups)
```
Comment (yogsototh) [inline src/ctia/http/middleware/ratelimit.clj]:
```suggestion
                         auth/groups
```

File: src/ctia/http/middleware/ratelimit.clj [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            (catch ExceptionInfo e
+              (let [origin (-> (ex-data e) :origin)]
+                (if (= origin :ring-turnstile-middleware)
+                  (do
+                    (log/error e "Unexpected error in the rate limit middleware")
+                    (handler request))
+                  (throw e)))))))
+      (do
+        (log/warn "Rate limit is disabled")
+        (fn [request] (handler request))))))
```
Comment (yogsototh) [inline src/ctia/http/middleware/ratelimit.clj]:
```suggestion
        handler))))
```

File: src/ctia/http/middleware/ratelimit.clj [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+                 :rate-limit-handler rate-limited-request-handler
+                 :key-prefix key-prefix}))]
+        (fn [request]
+          (try
+            (turnstile-mw request)
+            (catch ExceptionInfo e
+              (let [origin (-> (ex-data e) :origin)]
+                (if (= origin :ring-turnstile-middleware)
+                  (do
+                    (log/error e "Unexpected error in the rate limit middleware")
```
Comment (yogsototh) [inline src/ctia/http/middleware/ratelimit.clj]:
```suggestion
                    (log/errorf "Unexpected error in the rate limit middleware: %s" (pr-str e))
```

I'm not 100% certain about that one.

File: README.md (hunk 1 above)
Comment (msprunck) [reply README.md]: `X-RateLimit-GROUP-Remaining` is not returned

File: test/ctia/http/middleware/ratelimit_test.clj (hunk 2 above)
Comment (msprunck) [reply test/ctia/http/middleware/ratelimit_test.clj]: `X-RateLimit-GROUP-Remaining` is not returned

**Comment by hshahman** [comment]
Currently, CTR UI users are impacted by the CTIA rate limiting. 
When UI users send query to CTIA, the response header contains `x-ratelimit-group-limit": "8000"`, which is not the expected behavior.

**Comment by hshahman** [comment]
Verified in rel-v1.29

**Pull request merged**